### PR TITLE
Added InferDictionariesFromRecords flag to JSON TP

### PR DESCRIPTION
**Pull request by nikoyak**
JSON TP always projects json record fields into the properties
of generated types. Sometimes it is not desirable. The added
flag will allow to infer dictionaries from the records if the names
of all the fields of similar records in the provider's training data
are inferred to the same non-string primitive type (by type inference
rules). The flag is not set by default for backwards compatibility.

The current implementation makes it meaningless to process such data in the JSON Provider:
```json
{ "sales": { "2022-01-13": { "count": 3 }, "2022-01-23": { "count": 5 }, "2022-01-28": { "count": 4 }, "2022-02-01": { "count": 5 }, "2022-02-07": { "count": 1 }, "2022-02-13": { "count": 2 } } }
```
With this PR, for example, it is easy to find out from the above data how many sales were in February:
```f#
type TP = JsonProvider<"""{ "sales": { "1970-01-01": { "count": 100 }}}""", InferDictionariesFromRecords = true>
let s = """{ "sales": { "2022-01-13": { "count": 3 }, "2022-01-23": { "count": 5 }, "2022-01-28": { "count": 4 }, "2022-02-01": { "count": 5 }, "2022-02-07": { "count": 1 }, "2022-02-13": { "count": 2 } } }"""
TP.Parse(s).Sales.Items
|> Seq.filter (fun (k, _) -> k.Month = 2)
|> Seq.sumBy (fun (_, v) -> v.Count)
```
In addition to the `Items: (InferedKey * InferedValue) seq` property specified in the code above, the generated dictionary-like type contains the `Count`, `IsEmpty`, `Keys`, `Values` properties, the `Item` indexed property, and the `ContainsKey`, `TryFind` methods, similar to the members of `FSharp.Collections.Map`

### CODE REVIEW

File: src/Json/JsonGenerator.fs [hunk 1]
```diff
@@ -291,8 +294,94 @@ module JsonTypeBuilder =
...
-        // Add all record fields as properties
-        let members = 
+        let inferedKeyValueType =
+          let aggr = List.fold (StructuralInference.subtypeInfered false) InferedType.Top
+          let dropRecordName infType =
+            match infType with
+            | InferedType.Record (_, fields, opt) -> InferedType.Record (None, fields, opt)
+            | _  -> infType
+
+          if not <| ctx.InferDictionariesFromRecords
```
[the remark this answers is not in the document]
Comment (nikoyak) [reply src/Json/JsonGenerator.fs:304]: of course

**Comment by dsyme** [comment]
Suggest `PreferDictionaries`  instead of `InferDictionariesFromRecords`

File: src/Json/JsonGenerator.fs [hunk 2]
```diff
@@ -291,8 +294,91 @@ module JsonTypeBuilder =
...
+
+          let keysGetter = fun (Singleton jDoc) -> 
+            ctx.JsonRuntimeType?GetKeysFromInferedDictionary (keyResult.ConvertedType) (jDoc, keyResult.ConverterFunc ctx)
+
+          let valuesGetter = fun (Singleton jDoc) -> 
+            ctx.JsonRuntimeType?GetValuesFromInferedDictionary (valueConvertedTypeErased) (jDoc, valueResult.ConverterFunc ctx)
+
+          let (|Doubleton|) = function [f; s] -> f, s | _ -> failwith "Parameter mismatch"
+          
+          let itemGetter = fun (Doubleton (jDoc, key)) ->
```
Comment (dsyme) [inline src/Json/JsonGenerator.fs:339]: Define these as functions

**Review by dsyme — APPROVED** [review]

**Comment by dsyme** [comment]
I left a couple of suggestions, but otherwise marked as approved, the code looks great

**Comment by nikoyak** [comment]
done

**Review by cartermp — APPROVED** [review]

**Comment by cartermp** [comment]
Thanks!

**Pull request merged**